torch/utils/checkpoint.py

```diff
@@ -335,7 +339,275 @@ def forward(input):
+# 2) Pack the tensor as-is so that one may perform backward on the recomputed
+#    graph. The tensors saved to this graph will live until the end of
+#    recomputation, or earlier if someone performs backward with
+#    retain_graph=False or something.
+#
+# [ Multiple backwards ]
+#
+# The example below shows what happens if during recomputation we find that some
+# of the tensors we are trying to recompute have already been cleared.
+#
+# Spoiler: we don't do anything special, we just skip over them!
+#
+# def fn(x):
+#   y = x.sin()                           # (1)
+#   z = y.cos()                           # (2)
+#   gx, = torch.autograd.grad(z, x)       # (3)
+#   w = x.sin()                           # (4)
+#   v = w.cos()                           # (5)
+#   gx2, = torch.autograd.grad(v, x)      # (6)
+#   return x * gx * gx2
+#
+# out = checkpoint(fn)(inp)
+#
+# In the code above fn is computed (potentially partially) 4 times in total.
+#   1. Don't save x and y since we are inside a checkpoint.
+#   2. Trigger a recompute of fn as we reach (3) since x and y weren't saved.
+#   3. If early stop is enabled, stop at (2)
+#   4. Continue original forward at (4), not saving x and w.
+#   5. (5) triggers a recompute of fn
+#   6. During recompute, we see that in the original graph, gx has already
+#      cleared x and y since backward is run at (3) without retain_graph=True
+#      We save x and w, however.
+#   7. Continue with returning
+#
+
+# NB: This is temporary and should be removed in a follow up PR. Early stopping
+#     is currently disabled by default. Since some nested test cases require
+#     ealry stopping to pass, _set_checkpoint_early_stop can be used to enable.
+_enable_checkpoint_early_stop = False
+
+@contextlib.contextmanager
+def set_checkpoint_early_stop(enable):
+    global _enable_checkpoint_early_stop
+    try:
+        prev = _enable_checkpoint_early_stop
+        _enable_checkpoint_early_stop = enable
+        yield
+    finally:
+        _enable_checkpoint_early_stop = prev
+
+class _Handle():
+    pass
+
+class _Holder():
+    def __init__(self):
+        self.handles: Dict[int, Optional[_Handle]] = dict()
+
+class _NoopSaveInputs(torch.autograd.Function):
+    @staticmethod
+    def forward(*args):
+        return torch.empty((0,))
+
+    @staticmethod
+    def setup_context(ctx: Any, inputs: Tuple[Any, ...], output: Any) -> None:
+        # Only tensors can be saved with ctx.save_for_backward, everything else
+        # is captured by get_args, which is saved directly on ctx
+        tensor_indices, tensors = zip(*[(i, o) for i, o in enumerate(inputs) if isinstance(o, torch.Tensor)])
+        idx2saved_idx = {b: a for a, b in enumerate(tensor_indices)}
+        args = [None if isinstance(o, torch.Tensor) else o for o in inputs]
+
+        def get_args(saved_tensors):
+            ret = [saved_tensors[idx2saved_idx[i]] if i in tensor_indices else o for i, o in enumerate(args)]
+            return ret[1:]
```

Comment:
That sounds ok but feels quite complicated. Two for loops would be easier to read than all these complex list comprehensions I think. But if you prefer this, that's fine.

Reply:
Keeping this but beefing up the comments here a little